src/cargo/util/toml/mod.rs

```diff
@@ -368,6 +368,42 @@ pub fn prepare_for_publish(
         lints: me.lints.clone(),
     });
 
+    fn map_feature(
+        features: Option<&BTreeMap<manifest::FeatureName, Vec<String>>>,
+        deps: Option<&BTreeMap<manifest::PackageName, manifest::InheritableDependency>>,
+    ) -> CargoResult<Option<BTreeMap<manifest::FeatureName, Vec<String>>>> {
+        let Some(features) = features else {
+            return Ok(None);
+        };
+        let features = features
+            .iter()
+            .map(|(name, feature_deps)| {
+                let feature_deps = feature_deps
+                    .iter()
+                    .filter(|feature_dep| {
+                        if let Some(feature_pkg_name) = feature_dep.split('/').next() {
+                            let Some(deps) = deps else { return false };
+                            let k =
+                                &manifest::PackageName::new(feature_pkg_name.to_string()).unwrap();
+                            match deps.get(k) {
+                                Some(manifest::InheritableDependency::Value(def)) => {
+                                    def.is_version_specified()
+                                }
+                                _ => false,
```

Comment:
Here we've got the same symptom mentioned in this comment <https://github.com/rust-lang/cargo/pull/12230#issuecomment-1575658799>:

* [Platform specific dependencies](https://doc.rust-lang.org/nightly/cargo/reference/specifying-dependencies.html#platform-specific-dependencies) (`target.<…>.dev-dependencies`) should be handled.
* A dependency might appear in both `[dev-dependencies]` other kind of dependency table, and we don't want to remove those appear in `[dependencies]` or [`build-dependencies`] because they are essential.

A good way to ensure that we don't introduce any bug is to write some tests to verify and record the current “incorrect” behavior in the first commit. Then the following commits fix both the issue and tests. This also help reviewer to follow the workflow.

Reply:
Tests now cover every combination I can think of